Introduce the KCC provider for servicemapping and sample generation in Config Connector

maqiuyujoyce opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->
Add a Terraform KCC provider to generate KCC servicemappings and TF samples automatically.

Used the Terraform provider to reuse the existing logic that reads and parses the api.yaml files and the additional terraform.yaml layer.

Context: b/258035442

<!--
Replace each [ ] with [X] to check it. Switch to the preview view to make it easier to click on links.
These steps will speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.
-->
If this PR is for Terraform, I acknowledge that I have:

- [ ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ ] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I've detected that you're a community contributor. @megan07, a repository maintainer, has been assigned to assist you and help review your changes. 

<details>
  <summary>:question: First time contributing? Click here for more details</summary>

---

Your assigned reviewer will help review your code by: 
* Ensuring it's backwards compatible, covers common error cases, etc.
* Summarizing the change into a user-facing changelog note.
* Passes tests, either our "VCR" suite, a set of presubmit tests, or with manual test runs.

You can help make sure that review is quick by running local tests and ensuring they're passing in between each push you make to your PR's branch. Also, try to leave a comment with each push you make, as pushes generally don't generate emails.

If your reviewer doesn't get back to you within a week after your most recent change, please feel free to leave a comment on the issue asking them to take a look! In the absence of a dedicated review dashboard most maintainers manage their pending reviews through email, and those will sometimes get lost in their inbox.

---

</details>

maqiuyujoyce commented:
I didn't check any checkbox as this change is not for the Terraform provider. Let me know if there is anything else I should do about this PR.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

rileykarson commented on mmv1/provider/terraform_kcc.rb:
```diff
@@ -0,0 +1,162 @@
+# Copyright 2019 Google Inc.
```
```suggestion
# Copyright 2023 Google Inc.
```

maqiuyujoyce replied:
Good catch! Done.

rileykarson commented on mmv1/provider/terraform_kcc.rb:
```diff
@@ -0,0 +1,162 @@
+
+    def generate_resource_sweepers(pwd, data) end
+
+    def generate_iam_policy(pwd, data, generate_code, generate_docs)end
```
```suggestion
    def generate_iam_policy(pwd, data, generate_code, generate_docs) end
```

maqiuyujoyce replied:
Done.

rileykarson reviewed (approved):
Looks reasonable to me! I'm not too familiar with the modern ServiceMapping format so I can't validate that very well, but the parts I am familiar with make sense. Two minor nitpicks (a copyright and spacing), let me know if you'd rather I just merge this rather than wait on them getting updated.

maqiuyujoyce commented:
Thank you, @rileykarson ! Addressed the comments and updated the template to use the current year for copyright (instead of using a static year).

> I'm not too familiar with the modern ServiceMapping format so I can't validate that very well, but the parts I am familiar with make sense.

Thanks! We'll probably want to rely on the tests in config connector to do the validation.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

modular-magician commented:
#### Tests analytics 
Total tests: `2405` 
Passed tests `2151` 
Skipped tests: `251` 
Failed tests: `3` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccContainerCluster_withInvalidGatewayApiConfigChannel|TestAccBillingBudget_billingBudgetFilterExample|TestAccBillingBudget_billingBudgetUpdate </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccContainerCluster_withInvalidGatewayApiConfigChannel`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/e33d69c4-393a-49d2-bbe1-8158c810e813/recording/TestAccContainerCluster_withInvalidGatewayApiConfigChannel.log)]
`TestAccBillingBudget_billingBudgetFilterExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/e33d69c4-393a-49d2-bbe1-8158c810e813/recording/TestAccBillingBudget_billingBudgetFilterExample.log)]
`TestAccBillingBudget_billingBudgetUpdate`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/e33d69c4-393a-49d2-bbe1-8158c810e813/recording/TestAccBillingBudget_billingBudgetUpdate.log)] 

All tests passed
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/e33d69c4-393a-49d2-bbe1-8158c810e813/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/e33d69c4-393a-49d2-bbe1-8158c810e813/recording) for each test

rileykarson commented on mmv1/templates/kcc/product/service_mapping.yaml.erb:
```diff
@@ -0,0 +1,132 @@
+<%
+autogen_exception
+-%>
+# Copyright <%= Time.new.year %> Google LLC
```
```suggestion
# Copyright 2023 Google LLC
```

rileykarson commented on mmv1/templates/kcc/samples/sample.tf.erb:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Copyright <%= Time.new.year %> Google LLC
```
```suggestion
 * Copyright 2023 Google LLC
```

rileykarson reviewed (commented):
Ah, some unit tests are complaining about the generated copyright dates. Let's just set static current ones, that should be fine.

We should be able to ignore the CLA check failure, since Google's the owner of your change anyways. I'll override.

rileykarson reviewed (approved):
Oh! The issue is a little different than I thought. We've got a hardcoded list of date formats in https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/spec/copyright.rb, and `23` was only recently added as a suffix. In your branch (https://github.com/maqiuyujoyce/magic-modules/blob/master/mmv1/spec/copyright.rb) it's not included yet. While this fails right now, merging will resolve it.

LGTM

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR hasn't generated any diffs, but I'll let you know if a future commit does.

modular-magician commented:
#### Tests analytics 
Total tests: `2407` 
Passed tests `2152` 
Skipped tests: `251` 
Failed tests: `4` 

#### Action taken 
<details> <summary>Triggering VCR tests in RECORDING mode for the tests that failed during VCR. Click here to see the failed tests</summary><blockquote>TestAccFirebaserulesRelease_BasicRelease|TestAccRegionInstanceGroupManager_stateful|TestAccBillingBudget_billingBudgetUpdate|TestAccBillingBudget_billingBudgetFilterExample </blockquote></details>

modular-magician commented:
Tests passed during RECORDING mode:
 `TestAccFirebaserulesRelease_BasicRelease`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/dd8e7743-0402-4c65-89f3-bbb8e9b7bff1/recording/TestAccFirebaserulesRelease_BasicRelease.log)]
`TestAccBillingBudget_billingBudgetUpdate`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/dd8e7743-0402-4c65-89f3-bbb8e9b7bff1/recording/TestAccBillingBudget_billingBudgetUpdate.log)]
`TestAccBillingBudget_billingBudgetFilterExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/dd8e7743-0402-4c65-89f3-bbb8e9b7bff1/recording/TestAccBillingBudget_billingBudgetFilterExample.log)] 

Tests failed during RECORDING mode:
 `TestAccRegionInstanceGroupManager_stateful`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/dd8e7743-0402-4c65-89f3-bbb8e9b7bff1/build-log/recording_build/TestAccRegionInstanceGroupManager_stateful_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/dd8e7743-0402-4c65-89f3-bbb8e9b7bff1/recording/TestAccRegionInstanceGroupManager_stateful.log)] 

Please fix these to complete your PR
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/dd8e7743-0402-4c65-89f3-bbb8e9b7bff1/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-6970/artifacts/dd8e7743-0402-4c65-89f3-bbb8e9b7bff1/recording) for each test
